sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/PruneShuffleAndSort.scala

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.exchange
+
+import org.apache.spark.sql.catalyst.expressions.SortOrder
+import org.apache.spark.sql.catalyst.plans.physical.{HashPartitioning, PartitioningCollection}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{SortExec, SparkPlan}
+
+/**
+ * Removes unnecessary shuffles and sorts after new ones are introduced by [[Rule]]s for
+ * [[SparkPlan]]s, such as [[EnsureRequirements]].
+ */
+case class PruneShuffleAndSort() extends Rule[SparkPlan] {
+
+  override def apply(plan: SparkPlan): SparkPlan = {
+    plan.transformUp {
+      case operator @ ShuffleExchangeExec(upper: HashPartitioning, child, _) =>
+        child.outputPartitioning match {
+          case lower: HashPartitioning if upper.semanticEquals(lower) => child
```

Comment:
Why doesn't this apply to `RangeParititoning`? Is it just we assume `repartition()` would only do `HashPartitioning`?
And what happens if someone does `df1.join(df2, Seq("id"), "left").repartition(100, df("some_other_column")).repartition(20, df1("id"))` or `df1.join(df2, Seq("id"), "left").sortWithinPartition(df1("some_other_column")).sortWithinPartition(df1("id"))` ? We should be able to optimize that out too, right? It would be nice to make this rule more general and cover a wider range of cases.

Reply:
Thanks for providing feedback.
Let me take a look into your specific examples  and think a little more about it.

Reply:
This PR focused on fixing removing unnecessary sorting and shuffling after a join, which potentially includes its own ShuffleExchangeExec with HashPartioning. Both cases you mentioned are already optimized properly:  the shuffling on "some_other_column" is removed and all sortWithinPartitions are removed (due to previous optimizations in logical plan, and the optimizations introduced here)

I wouldn't mind generalizing to all Partitioning types of the ShuffleExchangeExec, but I am not sure how to compare two partitioning types for equality. You can see the special case for HashPartitioning in this PR.

Reply:
@maryannxue ping